File: src/main/java/io/github/takenoha/towerdefense/runtime/EnemyBridgePlan.java
package io.github.takenoha.towerdefense.runtime;

import java.util.Locale;
import java.util.Objects;

/** One conservative, one-block builder placement selected from a world snapshot. */
public record EnemyBridgePlan(String targetMaterialKey) {
    public EnemyBridgePlan {
        Objects.requireNonNull(targetMaterialKey, "targetMaterialKey");
        if (targetMaterialKey.isBlank()) {
            throw new IllegalArgumentException("targetMaterialKey must not be blank");
        }
        String normalized = targetMaterialKey.toLowerCase(Locale.ROOT);
        if (normalized.equals("minecraft:air")
                || normalized.equals("minecraft:cave_air")
                || normalized.equals("minecraft:void_air")) {
            throw new IllegalArgumentException("a bridge target must not be air");
        }
        if (TerrainMutationPolicy.isRequiredMaterial(targetMaterialKey)) {
            throw new IllegalArgumentException(
                    "a bridge target must not be a required protected material");
        }
    }
}

File: src/main/kotlin/io/github/takenoha/towerdefense/runtime/EnemyBridgePlan.kt
package io.github.takenoha.towerdefense.runtime

import java.util.Locale
import java.util.Objects
import kotlin.jvm.JvmRecord

/** One conservative, one-block builder placement selected from a world snapshot. */
@JvmRecord
data class EnemyBridgePlan(
    val targetMaterialKey: String,
) {
    init {
        Objects.requireNonNull(targetMaterialKey, "targetMaterialKey")
        if (targetMaterialKey.isBlank()) {
            throw IllegalArgumentException("targetMaterialKey must not be blank")
        }
        val normalized = targetMaterialKey.lowercase(Locale.ROOT)
        if (normalized == "minecraft:air" ||
            normalized == "minecraft:cave_air" ||
            normalized == "minecraft:void_air"
        ) {
            throw IllegalArgumentException("a bridge target must not be air")
        }
        if (TerrainMutationPolicy.isRequiredMaterial(targetMaterialKey)) {
            throw IllegalArgumentException(
                "a bridge target must not be a required protected material",
            )
        }
    }
}
